### PR TITLE
Configure a Sentinel Connection when initialising the Sentinel pool

### LINKED ISSUE #1487: Timeout settings not applied to sentinel connection
### Expected behavior

The `connectionTimeout` and `soTimeout` arguments in `JedisSentinelPool` constructor are applicable to all connections (connection to Sentinels, and Master).

### Actual behavior

The connection timeout of Sentinel connection is 2000 ms regardless of `connectionTimeout` and `soTimeout` value. Is this purposely designed or a bug?

### Steps to reproduce:

1. Prepare non-routable address for timeout simulation, e.g. 10.0.0.0:26379
2. Set the connection timeout to 10000 ms
3. `new JedisSentinelPool(masterName, nonRoutable, config, connectionTimeout)`
4. Observe the actual timeout, it will happen after 2000 ms instead of 10000 ms

### PR DESCRIPTION
**Pull request by karltinawi**
This commit attempts to solve #1487 and #1636.

This PR adds a constructor to JedisSentinelPool to configure a Sentinel connection with respective timeout settings and a password. Supporting constructors have also been added to BinaryJedis and Jedis classes.

Thanks in advance for your review.

### CODE REVIEW

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 1]
```diff
@@ -86,10 +89,22 @@ public JedisSentinelPool(String masterName, Set<String> sentinels,
...
       final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout,
       final String password, final int database, final String clientName) {
+    this(masterName, sentinels, poolConfig, connectionTimeout, soTimeout, password, Protocol.DEFAULT_TIMEOUT,
+        Protocol.DEFAULT_TIMEOUT, Protocol.DEFAULT_PASSWORD, database, clientName);
+  }
+
+  public JedisSentinelPool(String masterName, Set<String> sentinels,
+      final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout, final String password,
+      final int sentinelConnectionTimeout, final int sentinelSoTimeout, final String sentinelPassword,
+      final int database, final String clientName) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Add all sentinel parameters after all master parameters, i.e. after `clientName`.

File: src/main/java/redis/clients/jedis/Protocol.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
   private static final String BUSY_PREFIX = "BUSY ";
   private static final String NOSCRIPT_PREFIX = "NOSCRIPT ";
 
+  public static final String DEFAULT_PASSWORD = null;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Protocol.java]: Remove default password variable.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 3]
```diff
@@ -146,7 +161,12 @@ private HostAndPort initSentinels(Set<String> sentinels, final String masterName
 
       Jedis jedis = null;
       try {
-        jedis = new Jedis(hap);
+        // Being explicit here, although this null check is likely not required as the underlying client ignores 'null' password anyway.
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Agree. Just keep one.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 4]
```diff
@@ -84,14 +87,26 @@ public JedisSentinelPool(String masterName, Set<String> sentinels,
...
+      final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout, final String password,
+	   final int database, final String clientName) {
+    this(masterName, sentinels, poolConfig, connectionTimeout, soTimeout, password, database, clientName,
+        Protocol.DEFAULT_TIMEOUT, Protocol.DEFAULT_TIMEOUT, null);
+  }
+
+  public JedisSentinelPool(String masterName, Set<String> sentinels,
+      final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout, final String password,
+       final int database, final String clientName,
+       final int sentinelConnectionTimeout, final int sentinelSoTimeout, final String sentinelPassword) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: I think, after this, a request for sentinelClientName would come right away. IMO, we can add it now.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 5]
```diff
@@ -146,7 +161,7 @@ private HostAndPort initSentinels(Set<String> sentinels, final String masterName
 
       Jedis jedis = null;
       try {
-        jedis = new Jedis(hap);
+        jedis = new Jedis(hap.getHost(), hap.getPort(), sentinelConnectionTimeout, sentinelSoTimeout, sentinelPassword);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Without sending password and clientName in constructor, auth and clientSetName can be done here.

File: src/main/java/redis/clients/jedis/BinaryJedis.java [hunk 6]
```diff
@@ -95,6 +95,14 @@ public BinaryJedis(final String host, final int port, final int connectionTimeou
     client.setSoTimeout(soTimeout);
   }
 
+  public BinaryJedis(final String host, final int port, final int connectionTimeout,
+      final int soTimeout, final String password) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedis.java]: There isn't any other constructor with direct password param. The constructor with URI accepts password and does auth. I'm not so sure what would be better here also considering inclusion of client name. IMHO, remove this constructor for the time being.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 7]
```diff
@@ -78,6 +78,11 @@ public Jedis(final String host, final int port, final int connectionTimeout, fin
     super(host, port, connectionTimeout, soTimeout);
   }
 
+  public Jedis(final String host, final int port, final int connectionTimeout, final int soTimeout,
+      final String password) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Remove this constructor. See comment in BinaryJedis.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 4 above)
Comment (karltinawi) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Agreed. I'll add an alternative constructor, similar to initialising the pool with/without a client name. Unless you disagree?

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 5 above)
Comment (karltinawi) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: True. I'll make that change.

Comment (karltinawi) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: I've followed the same approach as `JedisFactory` in configuring Jedis for connecting to sentinel (i.e. when respective value is not null).

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 8]
```diff
@@ -84,14 +88,36 @@ public JedisSentinelPool(String masterName, Set<String> sentinels,
...
   public JedisSentinelPool(String masterName, Set<String> sentinels,
-      final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout,
-      final String password, final int database, final String clientName) {
+      final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout, final String password,
+      final int database, final String clientName) {
+    this(masterName, sentinels, poolConfig, connectionTimeout, soTimeout, password, database, clientName,
+        Protocol.DEFAULT_TIMEOUT, Protocol.DEFAULT_TIMEOUT, null);
+  }
+
+  public JedisSentinelPool(String masterName, Set<String> sentinels,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Remove this constructor. Let's keep only the later one.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 9]
```diff
@@ -146,7 +172,13 @@ private HostAndPort initSentinels(Set<String> sentinels, final String masterName
 
       Jedis jedis = null;
       try {
-        jedis = new Jedis(hap);
+        jedis = new Jedis(hap.getHost(), hap.getPort(), sentinelConnectionTimeout, sentinelSoTimeout);
+        if (sentinelPassword != null) {
+          jedis.auth(sentinelPassword);
+		}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Format this and later lines properly.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 10]
```diff
@@ -17,14 +17,18 @@
...
 
-  protected int connectionTimeout = Protocol.DEFAULT_TIMEOUT;
-  protected int soTimeout = Protocol.DEFAULT_TIMEOUT;
-
+  protected int connectionTimeout;
+  protected int soTimeout;
   protected String password;
 
-  protected int database = Protocol.DEFAULT_DATABASE;
+  protected int sentinelConnectionTimeout;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Nit-picky, but keep all master variables before all sentinel variables.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 4 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Let's just do it in one constructor.

**Comment by karltinawi** [comment]
Thanks @sazzad16.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 11]
```diff
@@ -17,15 +17,20 @@
...
 
-  protected int connectionTimeout = Protocol.DEFAULT_TIMEOUT;
-  protected int soTimeout = Protocol.DEFAULT_TIMEOUT;
-
+  protected int connectionTimeout;
+  protected int soTimeout;
   protected String password;
 
-  protected int database = Protocol.DEFAULT_DATABASE;
+  protected int database;
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: final?

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 12]
```diff
@@ -17,15 +17,20 @@
...
 
   protected String clientName;
 
+  protected int sentinelConnectionTimeout;
+  protected int sentinelSoTimeout;
```
Comment (gkorland) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java]: should all these fields be final?

**Comment by silentFred** [comment]
What is the holdup on this PR? Would be great to have this sometime

**Pull request merged**